src/hitchhiker/tree/core.clj

```diff
@@ -301,174 +390,202 @@
           path
           (recur (pop tmp)))))))
 
+
 (defn right-successor
   "Given a node on a path, find's that node's right successor node"
   [path]
   ;(clojure.pprint/pprint path)
   ;TODO this function would benefit from a prefetching hint
   ;     to keep the next several sibs in mem
-  (when-let [common-parent-path
-             (backtrack-up-path-until
-               path
-               (fn [parent index]
-                 (< (inc index) (count (:children parent)))))]
-    (let [next-index (-> common-parent-path peek inc)
-          parent (-> common-parent-path pop peek)
-          new-sibling (resolve (nth (:children parent) next-index))
-          ;; We must get back down to the data node
-          sibling-lineage (into []
-                                (take-while #(or (index-node? %)
-                                                 (data-node? %)))
-                                (iterate #(let [c (-> % :children first)]
-                                            (if (tree-node? c)
-                                              (resolve c)
-                                              c))
-                                         new-sibling))
-          path-suffix (-> (interleave sibling-lineage
-                                      (repeat 0))
-                          (butlast)) ; butlast ensures we end w/ node
-          ]
-      (-> (pop common-parent-path)
-          (conj next-index)
-          (into path-suffix)))))
-
-(defn forward-iterator
-  "Takes the result of a search and returns an iterator going
+  (go-try
+    (when-let [common-parent-path
+               (backtrack-up-path-until
+                path
+                (fn [parent index]
+                  (< (inc index) (count (:children parent)))))]
+      (let [next-index (-> common-parent-path peek inc)
+            parent (-> common-parent-path pop peek)
+            new-sibling (<? (resolve (nth (:children parent) next-index)))
+            ;; We must get back down to the data node
+            sibling-lineage (loop [res [new-sibling]
+                                    s new-sibling]
+                              (let [c (-> s :children first)
+                                    c (if (tree-node? c)
+                                        (<? (resolve c))
+                                        c)]
+                                (if (or (index-node? c)
+                                        (data-node? c))
+                                  (recur (conj res c) c)
+                                  res)))
+            ;; iterate cannot do blocking operations with core.async
```

Comment:
Yes, otherwise I have to wrap go around function boundaries. So I opt out for loops unless they become to bloated. I also have explored a more declarative `go-for` macro for this case: https://github.com/replikativ/superv.async

Sadly it has some limitations and is not as fast (it also has to do nested error-handling and automatically wrap and unwrap everything in channels). I couldn't figure out doing it with transducers so far.